### ISSUE TITLE
Unexpected temporary attribute transition behaviour

### ISSUE THREAD

**Issue by mal** [renpy/renpy#1761]
Using `with dissolve` on a say line with a temporary attribute only affects the initial change, not the restoration. Conversely using `with dissolve` alone on a line following a say line with a temporary attribute only affects the restorative change.

In addition to this being strange, it also differs from behaviour of normal say lines which are un-impacted by `with dissolve` statements on separate lines.

```renpy
image r = Solid('f773', xysize=(320, 240))
image b = Solid('7ff3', xysize=(320, 240))
image y = Solid('ff73', xysize=(320, 240))

define foo = Character('Foo', image='square')

layeredimage square:
  group colour:
    attribute red default "r"
    attribute blue "b"
    attribute yellow "y"


label main_menu:
    $ _confirm_quit = False
    return


label start:
    scene expression '#111'
    show square at truecenter
    foo "start as red"
    foo @ blue "dissolve temporarily into blue" with dissolve
    foo "restore red (should dissolve back, but doesn't)"
    foo @ yellow "dissolve temporarily into yellow" with dissolve
    with dissolve
    foo "restore red (this time it {b}did{/b} dissolve back)"
    foo @ blue "temporarily switch to blue without dissolve"
    with dissolve
    foo "restore red (with dissolve once again)"

    foo yellow "this will dissolve to yellow" with dissolve
    with dissolve # this will act as a weird delay
    foo blue "this won't dissolve to blue"
    with dissolve # this will act as a weird delay

    "end"
    return
```

**Comment by renpytom** [comment]
I don't believe there's a problem here. The ``with`` clause for a say statement isn't meant as a shorthand for setting config.say_attribute_transition, which it kind of looks like you might think it is. Instead, it's shorthand for 

```
$ renpy.transition(dissolve)
"Say statement."
```

And as far as I can tell, it's doing that job flawlessly. Right now, there isn't syntax for setting a say_attribute_transition on a say statement basis, though say arguments and an appropriate function could work together to make this work. (A dedicated callback would be cleaner, though - I'd be happy to add that if you want.) 

I don't see this as a bug, however.

**Comment by mal** [comment]
Ah, OK that makes sense. The confusion came from seeing the transition apply when images were changing as a result of temporary attributes being added, but not they were changing due to them being reset.

Following your suggestion I did manage to make it work using say arguments, however the syntax change wasn't ideal (`"words" (with_=dissolve)`) as it would undoubtedly lead to "you need to put it in brackets, and use an equals and with needs to have an underscore on it"-type conversations with less technical team members, so I came up with the code below which instead steals the currently queued transition to use as the `config.say_attribute_transition` which seems to work for now.

Perhaps a feature request for "a say statement clause for setting config.say_attribute_transition for the duration of a single say statement" would be more appropriate? As the lack of transition on the attribute reset was the first thing I was asked about by a team member trying out the feature. It seems like the sort of thing that will come up again.

Thanks for pointing me in the right direction on this! 🙂

---

```renpy
init python hide:
    def say_arguments_callback(*args, **kwargs):
        config.say_attribute_transition = \
            renpy.get_transition(config.say_attribute_transition_layer)
        return (), kwargs

    config.say_arguments_callback = say_arguments_callback
```

**Comment by mal** [comment]
I had a look at the new `config.say_attribute_transition_callback` feature but it doesn't seem like it would be flexible enough to address the use case in this issue (I'm not really sure what use case it does address).

There's no way to signal it from a say statement, so it applies unilaterally, and it's only passed information about the `image_tag` which precludes the possibility of choosing a transition based upon anything but the layered image as a whole, rather than specific say statements as in the example, or even all say statements that cause a change in a specific group or attribute of a layered image.

Additionally because it fires a separate restore call the work around above using `renpy.get_transition` can't be leveraged within it.

**tl;dr**: while `config.say_attribute_transition_callback` may facilitate some use cases, it doesn't seem to resolve the use case faced here. Maybe closing this as "works as expected" and raising a more detailed and specific feature req is the way forward here?

**Comment by renpytom** [comment]
config.say_attribute_transition_callback is building block that should cover you. It's intended for use with  config.say_arguments_callback. The say arguments callback takes the transition, and stashes it somewhere. The config.say_attribute_transition_callback then retrieves and return it. 

This would let you write:

```
e happy "This is a test." (trans=dissolve)
```

To change the say attribute transition fore just one statement.

**Comment by mal** [comment]
Aha, that makes sense! Thanks!

Would it be possible to extend the callback slightly to receive a resolved list of tags/attributes from `images.apply_attributes` as well as `self.image_tag`? That would make it possible for the callback to dynamically choose a transition based upon the attributes being changed via something like:

```py
init python hide:
    dissolve_attrs = {'running', 'walking', 'jumping', 'skipping'}

    def example(tag, attrs, mode):
        current_attrs = _set(renpy.get_attributes(tag))
        future_attrs = _set(attrs)
        future_attrs.remove(tag)
        changed_attrs = current_attrs ^ future_attrs
        if dissolve_attrs & changed_attrs:
            return dissolve, config.say_attribute_transition_layer
        return config.say_attribute_transition, config.say_attribute_transition_layer

    config.say_attribute_transition_callback = example

```

Which would automatically use a dissolve transition anytime one of the preset attributes is added or removed.

---

Had a stab at a PoC for this, seems to work pretty well in my (admittedly limited) testing: https://github.com/renpy/renpy/pull/1774

**Comment by mal** [comment]
Closing this as it turns out that the unexpected behaviour is expected, and callback has been added to help in the situation encountered, with a PoC PR to further improve it.
